src/network/netdev/macsec.c

```diff
@@ -0,0 +1,911 @@
+                if (r < 0)
+                        return r;
+
+                r = netlink_call_async(netdev->manager->genl, NULL, m, macsec_add_tx_sa_handler,
+                                       netdev_destroy_callback, netdev);
+                if (r < 0)
+                        return log_netdev_error_errno(netdev, r, "Failed to configure secure association: %m");
+
+                netdev_ref(netdev);
+        }
+
+        ORDERED_HASHMAP_FOREACH(n, s->receive_association_by_section, i) {
+                r = netdev_macsec_receive_association_fill_message(netdev, n, &m);
+                if (r < 0)
+                        return r;
+
+                r = netlink_call_async(netdev->manager->genl, NULL, m, macsec_receive_association_handler,
+                                       netdev_destroy_callback, netdev);
+                if (r < 0)
+                        return log_netdev_error_errno(netdev, r, "Failed to configure receive association: %m");
+
+                netdev_ref(netdev);
+        }
+
+        return 0;
+}
+
+static int netdev_macsec_fill_message_create(NetDev *netdev, Link *link, sd_netlink_message *m) {
+        MACsec *v;
+        int r;
+
+        assert(netdev);
+        assert(m);
+
+        v = MACSEC(netdev);
+
+        r = sd_netlink_message_append_u16(m, IFLA_MACSEC_PORT, v->port);
+        if (r < 0)
+                return log_netdev_error_errno(netdev, r, "Could not append IFLA_MACSEC_PORT attribute: %m");
+
+        if (v->encrypt >= 0) {
+                r = sd_netlink_message_append_u8(m, IFLA_MACSEC_ENCRYPT, v->encrypt);
+                if (r < 0)
+                        return log_netdev_error_errno(netdev, r, "Could not append IFLA_MACSEC_ENCRYPT attribute: %m");
+        }
+
+        return r;
+}
+
+int config_parse_macsec_port(
+                const char *unit,
+                const char *filename,
+                unsigned line,
+                const char *section,
+                unsigned section_line,
+                const char *lvalue,
+                int ltype,
+                const char *rvalue,
+                void *data,
+                void *userdata) {
+
+        _cleanup_(macsec_receive_association_free_or_set_invalidp) ReceiveAssociation *b = NULL;
+        _cleanup_(macsec_receive_channel_free_or_set_invalidp) ReceiveChannel *c = NULL;
+        MACsec *s = userdata;
+        int r;
+
+        assert(filename);
+        assert(section);
+        assert(lvalue);
+        assert(rvalue);
+        assert(data);
+
+        if (streq(section, "MACsecReceiveChannel"))
+                r = macsec_receive_channel_new_static(s, filename, section_line, &c);
+        else
+                r = macsec_receive_association_new_static(s, filename, section_line, &b);
+
+        if (r < 0)
+                return r;
+
+        if (b)
+                r = parse_ip_port(rvalue, &b->rx.port);
+        else
+                r = parse_ip_port(rvalue, &c->port);
```

Comment:
is this actually an IP port? i.e. is 0 accepted here? parse_ip_port() filters that out... I mean, as I understand this is layer below IP, hence using an "IP port" parser is a bit of a layer violation, so it at least deserves a comment here, but I wonder what about port 0 here, does it have the same special meaning as on IP?

Reply:
The port used to make Secure Channel Identifier (SCI). ```0``` is a invalid value.